lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,393 @@
+      Insert(lowerVal, upperVal, source);
+    } else if (kind_ == sizeof(std::u32string::value_type)) {
+      const auto lowerVal =
+          lower ? GetStringValue<std::u32string>(lower->thing) : std::nullopt;
+      const auto upperVal =
+          upper ? GetStringValue<std::u32string>(upper->thing) : std::nullopt;
+      Insert(lowerVal, upperVal, source);
+    } else {
+      messages_.Say(source,
+          "Character kind type of case value (=%d) is not supported"_err_en_US,
+          kind_);
+    }
+  }
+  bool IsValidSelectCaseType(
+      const parser::ConstantExpr &constCase, parser::CharBlock src) {
+    if (const auto *caseval{GetExpr(constCase)}) {
+      if (auto type{caseval->GetType()}) {
+        if (type->category() != category_) {  // C1147 (R1140)
+          if (category_ == TypeCategory::Integer) {
+            messages_.Say(src, "CASE value must be of type INTEGER"_err_en_US);
+          } else if (category_ == TypeCategory::Character) {
+            messages_.Say(
+                src, "CASE value must be of type CHARACTER"_err_en_US);
+          } else {  // TypeCategory::Logical
+            messages_.Say(src, "CASE value must be of type LOGICAL"_err_en_US);
+          }
+        } else if (category_ == TypeCategory::Character) {
+          if (kind_ != type->kind()) {
+            messages_.Say(src,
+                "Character kind type of case construct (=%d) mismatches with the kind type of case value (=%d)"_err_en_US,
+                kind_, type->kind());
+          } else {
+            return true;
+          }
+        } else {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+private:
+  template<typename T> struct SelectCaseRangeType {
+    SelectCaseRangeType(std::optional<T> fromVal, std::optional<T> toVal,
+        parser::CharBlock source)
+      : fromVal(fromVal), toVal(toVal), source(source) {}
+    std::optional<T> fromVal;
+    std::optional<T> toVal;
+    parser::CharBlock source;
+  };
+  template<typename T>
+  bool IsMoreThanLowerVal(const T &strVal, const T &lowerStrVal) {
+    return (strVal.compare(lowerStrVal) >= 0);
+  }
+  bool IsMoreThanLowerVal(
+      const std::int64_t &intVal, const std::int64_t &lowerIntVal) {
+    return intVal >= lowerIntVal;
+  }
+  template<typename T>
+  bool IsLessThanUpperVal(const T &strVal, const T &upperStrVal) {
+    return (strVal.compare(upperStrVal) <= 0);
+  }
+  bool IsLessThanUpperVal(
+      const std::int64_t &intVal, const std::int64_t &upperIntVal) {
+    return intVal <= upperIntVal;
+  }
+  template<typename T>
+  bool CheckOverlappingValue(
+      const T &val, std::optional<T> lowerVal, std::optional<T> upperVal) {
+    return ((!lowerVal || IsMoreThanLowerVal(val, lowerVal.value())) &&
+        (!upperVal || IsLessThanUpperVal(val, upperVal.value())));
+  }
+  template<typename T>
+  bool CheckOverlappingRangeValue(std::optional<T> memberFromVal,
```

Comment:
This algorithm is overly complicated and the implementation is a mess.

Two intervals [a,b] and [c,d] overlap if and only if (c<=b && a<=d).

There's a utility Interval class template that used to be in lib/common/interval.h that might be helpful here.

Reply:
I will fix this as per your suggestion.